js/notifications.js

```diff
@@ -35,6 +35,11 @@
             if (this.length === 0) {
                 return;
             }
+            if (window.isFocused()) {
```

Comment:
I like that we're only doing this check in on place.

However, it does mean that after the 'empty' event, when the window is focused, every incoming unread message is added to the notifications collection then immediately removed. GIven that `conversation.notify()` does a `ConversationController.getOrCreateAndWait()`, perhaps it makes sense to add the `isFocused()` call back to prevent this extra unnecessary work?

Reply:
With that extra check, before the empty event, notifications won't be added to the collection if the window happens to be in focus when we process that message (possibly while the loading screen is still up). For a large backlog of messages that might mean an inaccurate count when we finally get the empty event and display the notification.

getOrCreateAndWait should be pretty lightweight given the initial contact fetch on startup, no?

Reply:
Good point. The "in-focus but not in-focus when we're finally ready to show the alert" is a key scenario I hadn't considered.

I think these changes are ready, just need a quick rebase and we can merge! :0)

Reply:
Ah, but there is one more wrinkle - in that case, where we're in focus now but we want to save the message for a later 'empty' event - that `windows.isFocused()` check will clear all current notifications. So we should probably remove the `clear()` call there.

Reply:
Well bail at the if !enabled check in that case.

Reply:
btw, I added the clear() there so that we don't amass a backlog of stale notifications for messages that arrive while the window is in focus. Otherwise, the next message to arrive while out of focus would include the backlog.